### DISCUSSION TITLE
CORS blocked for wavesurfer only, even if I use a media element

### DISCUSSION THREAD

**Discussion by gelform** [katspaugh/wavesurfer.js#3673] [Q&A]
**CORS is blocked for wavesurfer only, even if I use a media element, but the same URL works everywhere else.**

An mp3 is served from S3 via CloudFront. When I request the mp3 URL in the browser directly, the browser downloads the file. 

When I add an `<audio>` element to my page and use the URL as the source, the mp3 plays. 

However when I use the URL in the `url` param passed to wavesurfer, I get `has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

I found the thread suggesting using the `media` param: https://github.com/katspaugh/wavesurfer.js/discussions/3291

I tried both `document.createElement('audio')` and referencing the media element on the page that does play. I get the same error. 

wavesurfer successfully played MP3s from the same S3/CloudFront setup last week. No settings have been changed on AWS. 

wavesurfer is served locally so the version has not changed. The current version is `7.7.10`.

Finally, I removed all of my custom code and simply added the docs example to my page but got the same error:
```        <script src="https://unpkg.com/wavesurfer.js@7"></script>

        <script>

            const wavesurfer = WaveSurfer.create({
                container: '#waveform',
                waveColor: '#4F4A85',
                progressColor: '#383351',
                url: 'https://files.mydomain.com/my-file.mp3',
            })

            wavesurfer.on('interaction', () => {
                wavesurfer.play()
            })
        </script>
```

Am I doing something wrong? What am I missing? What else can I try? Thanks!

**Comment by katspaugh** [comment]
Please check the FAQ section of the readme, it’s covered there.

**Reply by gelform** [reply]
Sorry, I should've specified that I checked that as well as the other support threads.

I've confirmed that CORS is set correctly. The file works in native `<media>` elements. 

It **only** doesn't work in wavesurfer. That's why I'm unfortunately seeking guidance.

**Reply by katspaugh** [reply]
Media element doesn’t need CORS, wavesurfer does. You most likely did not set the headers correctly. Please Google or ask chatgpt how to do it with cloudfront.

**Comment by hieunm22** [comment]
did you find your solution.? I'm having same issue like you

**Comment by gabriel-inigo** [comment]
@gelform were you able to solve the problem?

**Reply by svalencias** [reply]
You can circumvent CORS in Javascript very easily, simply by loading the file, dealing with it via an arrayBuffer (with lowercase 'a') and then transforming it into an ArrayBuffer type inside AudioContext. After that you can serve it to wavesurfer.js or whatever is complaining about CORS at the time and it won´t complain anymore.

No need to change headers, have a local server or any of the usual malarkey. Just 6 very straightforward lines of vanilla JS and you can do local stuff without caring a iota about CORS. It´s way simpler than the code you could imagine you´d need to do something like this...and it works every time for me, no matter what file format you´re loading (if it´s already covered by AudioContext).

Yep, I hear you..."it´s impossible, right?" OK, download wavesurfer.min.js from a public repo, download the attached .TXT file as a .HTM file and paste it inside the same directory where you downloaded the wavesurfer.min.js library, pick up any audio file in your computer and see it for yourself.

Hope it helps. Enjoy!

Edited to update a version that deals with mono channel audio files and dual channel audio files seamlessly. Yesterday I added the created code by hand so as it was for explanatory purposes only, naturally I did not check on the small things that one can specify later but after thinking about it I think it would be what´s expected from a standard example and it´s good for you to have it from a 'completion' point of view anyway.

[NO_CORS_checking_channel_number.txt](https://github.com/user-attachments/files/20148530/NO_CORS_checking_channel_number.txt)

**Comment by entonbiba** [comment]
if its not coming from the same sub/domain and no CORS header is added to the header response all modern browsers will force it now. You can use a proxy server to stream from that would have the header response * to avoid this if you do not have access cloudfront to set the policy.

**Comment by lukemelnik** [comment]
Setting crossOrigin="anonymous" on my audio element seems to have fixed the problem for me
